freqtrade/tests/test_configuration.py

```diff
@@ -617,3 +617,86 @@ def test_validate_tsl(default_conf):
     default_conf['trailing_stop_positive_offset'] = 0.015
     Configuration(Namespace())
     configuration._validate_config_consistency(default_conf)
+
+
+# config['exchange'] subtree has required options in it
+# so it cannot be omitted in the config
+def test_load_config_default_exchange(all_conf) -> None:
+    del all_conf['exchange']
+
+    assert 'exchange' not in all_conf
+
+    with pytest.raises(ValidationError,
+                       match=r'\'exchange\' is a required property'):
+        configuration = Configuration(Namespace())
+        configuration._validate_config_schema(all_conf)
+###    assert 'exchange' in all_conf
```

Comment:
The point of this was to test if the whole sub-tree is added (hint: it can be - but only if the tree has a `default:{}` set).

It won't work for exchange the way we configured it - and it's a requirement to have it there since it has mandatory, non-default fields.

However it may work for telegram / webhooks... - and we should definitely add it for "experimental" - since that dictionary IS mandatory (but can be empty).
Side note: we shoudl then check to remove the code that's currently adding it - i belive it's in strategy-resolver ... but i'll maybe not in this PR

Although i'm not sure on the benefit of having "all" options / trees available if it's disabled, so it was an idea.

Reply:
I do not get you. Why any default value should be added if the exchange subtree was omitted? It's simply invalid -- since it contains a required option. If it's omitted, validation raises the Validation exception. Simply handle it and do not use that *non-valid* config. No sense to require if validation changes the config in this case and add some default value for omitted subtree - it's implementation specific. The implemenation of the validation may collect all default values for options and insert them into the config tree at the end of its work and only when the whole configuration is valid. Or may insert them separately for each validated option and subtree. It's not a matter at all since validation fails.

The benefit of having the all_config is not really a benefit but an approach to make tests working and working similar to each other. Simply because if we use default_config as in other tests, then the `del <option>` produces exception if <option> is not present in the default_config. That's why I added a separate config all_config which is to be all such options.

Reply:
I have it in the sample i posted as a sample for other config options.

Correct, it does not make sense for exchange - BUT it makes sense for "experimental". I posted the code as a sample - not as the ultimate test ... sorry if that was not clear


Deleting from `default_conf` should not cause other functions to fail - it's scope is function, so it'll be restored each time... 
If that really doesn't work, using default_conf.copy() before deleting would be another option ;)

Reply:
> Deleting from `default_conf` should not cause other functions to fail - it's scope is function, so it'll be restored each time...

I mean deletion from default_conf an option which is absent in it fails itself and produces exception. Change temporarily `all_conf` back to `default_conf` in a test for an option which is absent in default_conf (`'password'` or `'sandbox'`) and you'll see what I mean.

Reply:
ah yes, that's true, if it's not in default_conf to begin with then it'll fail, yes.